mcp/embedded_session_store.go

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2023-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package mcp
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost/server/public/model"
+)
+
+const embeddedSessionKeyPrefix = "mcp_embedded_session_id"
+
+func buildEmbeddedSessionKey(userID string) string {
+	return fmt.Sprintf("%s_%s", embeddedSessionKeyPrefix, userID)
+}
+
+// loadEmbeddedSessionID retrieves the stored embedded session ID for a user from KV
+// Returns empty string if none is stored
+func (m *ClientManager) loadEmbeddedSessionID(userID string) (string, error) {
+	key := buildEmbeddedSessionKey(userID)
+	var stored []byte
+	if err := m.pluginAPI.KV.Get(key, &stored); err != nil {
+		return "", fmt.Errorf("failed to retrieve embedded session from KV: %w", err)
+	}
+	return string(stored), nil
+}
+
+// storeEmbeddedSessionID stores the embedded session ID for a user in KV
+func (m *ClientManager) storeEmbeddedSessionID(userID, sessionID string) error {
+	key := buildEmbeddedSessionKey(userID)
+	if _, err := m.pluginAPI.KV.Set(key, []byte(sessionID)); err != nil {
+		return fmt.Errorf("failed to store embedded session in KV: %w", err)
+	}
+	return nil
+}
+
+func (m *ClientManager) deleteEmbeddedSessionID(userID string) error {
+	return m.pluginAPI.KV.Delete(buildEmbeddedSessionKey(userID))
+}
+
+// ensureEmbeddedSessionID ensures there is a valid embedded session for the user
+// It loads from KV, validates via Session.Get, and if missing/invalid, creates a new one
+// The created session is tagged for MCP via DeviceId and Props
+func (m *ClientManager) ensureEmbeddedSessionID(userID string) (string, error) {
+	if sessionID, err := m.tryReuseEmbeddedSession(userID); err != nil {
+		return "", err
+	} else if sessionID != "" {
+		return sessionID, nil
+	}
+
+	return m.createEmbeddedSession(userID)
+}
+
+func (m *ClientManager) tryReuseEmbeddedSession(userID string) (string, error) {
+	stored, err := m.loadEmbeddedSessionID(userID)
+	if err != nil {
+		m.log.Debug("Failed to load embedded session ID; will create new", "userID", userID, "error", err)
```

Comment:
The "will create new" feels presumptuous for a function that doesn't actually do that.